Add sampling functions

jisantuc opened the pull request:
This PR adds implementations for pre- and post-sampling energy statistics.

jisantuc commented:
Ok I think I have gotten to a point where I can say why maybe no one has done this before, and it has to do with hardware constraints.

On an Ubuntu laptop I have around the house with an [Intel Core i7-8565U CPU](https://ark.intel.com/content/www/us/en/ark/products/149091/intel-core-i78565u-processor-8m-cache-up-to-4-60-ghz.html), tests run successfully. In GitHub actions running tests with sudo**, I get "ERROR: MicroArchitecture not supported: 6A". On a c6a.large AWS EC2 instance, I get "ERROR: MicroArchitecture not supported: 1", and [AWS docs](https://aws.amazon.com/ec2/instance-types/c6a/) report "Amazon EC2 C6a instances are powered by 3rd generation AMD EPYC processors." Running on a [c6i](https://aws.amazon.com/ec2/instance-types/c6i/) ([all Intel Xeon 2020 or later which should have the RAPL interfaces available](https://ark.intel.com/content/www/us/en/ark/products/series/204098/3rd-generation-intel-xeon-scalable-processors.html)), I _also_ get a "ERROR: MicroArchitecture not supported: 6A". That message comes from [here](https://github.com/aservet1/jRAPL/blob/6a622d5a581b23b7653528fcd17f9e7a142d64fb/src/native/energy_check_utils.c#L127-L136), though I haven't traced through where that's called / what for yet.

I think this means we'd need a self-hosted runner in order to do this in CI (I can try out Azure / GCP if that's desired, but the hardware risk of "cloud provider chooses a different processor" is pretty big!). That adds its own wrinkles -- GitHub [says](https://docs.github.com/en/actions/hosting-your-own-runners/about-self-hosted-runners#self-hosted-runner-security) "We recommend that you only use self-hosted runners with private repositories," which dramatically curbs the potential impact this project could have. However, this is all ancillary to the technical challenge here, so I'll build the functions into sbt tasks and mark this PR ready for review.

** Tests need to be run with sudo to [deal with with a bunch of `pread error!` logs](https://github.com/aservet1/jRAPL/#prerequeisites), e.g. [in this run](https://github.com/47degrees/sbt-energymonitor/runs/5361859305?check_suite_focus=true)

juanpedromoreno commented on flake.lock:
```diff
@@ -0,0 +1,26 @@
+{
```
Does it make sense if we ignore this file? Sorry, my Nix knowledge is quite limited

jisantuc replied:
Huh, I might have misinterpreted a line in [this blog post](https://christine.website/blog/nix-flakes-2-2022-02-27) about what needs to be in source control. Testing...

jisantuc replied:
Yup `nix develop` (main thing I'm using it for right now, not actually building with it _yet_) isn't mad at all about gitignoring the lock file. It's removed in 127a13a / a872d05

juanpedromoreno reviewed (approved):
Thanks @jisantuc ! LGTM!

jisantuc commented:
Did policies change to re-require review on any changes? I don't remember this being the case previously but I don't know how much I've done in this GH org before. Anyway, changes since last review are really minimal, so should be good for a pretty cursory re-review

juanpedromoreno commented:
Hey @jisantuc , `.github` project also distributes the standard policies/settings like dismissing reviews after new commits.

jisantuc commented:
Maybe I just haven't added commits after a review before 🤔 🤷‍♂️
